### ISSUE TITLE
userfaultfd mem monitor implementation hangs on free()

### ISSUE THREAD

**Issue by jswaro** [ofiwg/libfabric#5554]
@shefty @swelch 

When running an MPI application with libfabric master, we see a hang in the application. This occurs with a 4.12.14.150.x SLE 15 SP0 kernel. Userfaultfd is advertising all features except the lowest bit (0x7e).  

The main application thread is stuck in kernel space, waiting for a completion for the event in the userfault FD code. The thread cannot be interrupted or woken up without a completion. I've added debug code to the application, and it is stuck in the syscall to free the memory associated with the sending buffer. The sending buffer is 512M in size. 

```
nid000003:~ # cat /proc/67421/stack
[<ffffffff8122fa81>] userfaultfd_event_wait_completion+0xc1/0x210
[<ffffffff81231c24>] userfaultfd_unmap_complete+0x84/0xd0
[<ffffffff81199117>] SyS_brk+0x1f7/0x220
[<ffffffff81002b05>] do_syscall_64+0x75/0xf0
[<ffffffff8160008f>] entry_SYSCALL_64_after_hwframe+0x42/0xb7
[<ffffffffffffffff>] 0xffffffffffffffff
```

The userfault FD thread is waiting on a lock in user space
```
nid000003:~ # cat /proc/67421/task/68463/stack
[<ffffffff810d1630>] futex_wait_queue_me+0xc0/0x120
[<ffffffff810d2813>] futex_wait+0x103/0x230
[<ffffffff810d4321>] do_futex+0x2d1/0xbe0
[<ffffffff810d4d5c>] SyS_futex+0x12c/0x160
[<ffffffff81002b05>] do_syscall_64+0x75/0xf0
[<ffffffff8160008f>] entry_SYSCALL_64_after_hwframe+0x42/0xb7
[<ffffffffffffffff>] 0xffffffffffffffff

(gdb) bt
#0  0x00002b33f1c40b9c in __lll_lock_wait_private () from /lib64/libc.so.6
#1  0x00002b33f1bbc639 in _int_free () from /lib64/libc.so.6
#2  0x00002b34143fac06 in mlx5_dereg_mr () from /usr/lib64/libmlx5-rdmav2.so
#3  0x00002b33f3a8b8fe in ibv_dereg_mr@@IBVERBS_1.1 () from /usr/lib64/libibverbs.so.1
#4  0x00002b33f1f837c5 in fi_ibv_mr_cache_delete_region (cache=0x45c068, entry=0x11498a60)
    at prov/verbs/src/verbs_mr.c:232
#5  0x00002b33f1f4748f in util_mr_free_entry (cache=0x45c068, entry=0x11498a60)
    at prov/util/src/util_mr_cache.c:81
#6  0x00002b33f1f4754d in util_mr_uncache_entry (cache=0x45c068, entry=0x11498a60)
    at prov/util/src/util_mr_cache.c:106
#7  0x00002b33f1f4760f in ofi_mr_cache_notify (cache=0x45c068, addr=0x414c5000, len=536870912)
    at prov/util/src/util_mr_cache.c:125
#8  0x00002b33f1f4499c in ofi_monitor_notify (monitor=0x2b33f2247120 <uffd>, addr=0x414c5000, len=536870912)
    at prov/util/src/util_mem_monitor.c:189
#9  0x00002b33f1f44e7e in ofi_uffd_handler (arg=0x2b33f2247120 <uffd>) at prov/util/src/util_mem_monitor.c:269
#10 0x00002b33f224f569 in start_thread () from /lib64/libpthread.so.0
#11 0x00002b33f1c33a2f in clone () from /lib64/libc.so.6
(gdb)
```

```
nid000003:~ # cat /proc/67421/maps | grep deleted
2b33ee077000-2b33ee24a000 rw-s 00000000 00:1b 37727185                   /dev/shm/shared_memory.PMI.4800.20 (deleted)
2b33f453a000-2b341417e000 rw-s 00000000 00:01 90669057                   /SYSV00000000 (deleted)
2b341418c000-2b341418d000 rw-s 00000000 00:1b 37730856                   /dev/shm/xyz (deleted)
2b341523b000-2b341526b000 rw-s 00000000 00:1b 37840341                   /dev/shm/abc (deleted)
2b3415400000-2b3415800000 rw-p 00000000 00:0c 37888074                   /anon_hugepage (deleted)
2b3415800000-2b3415c00000 rw-p 00000000 00:0c 37888115                   /anon_hugepage (deleted)
2b342d600000-2b342da00000 rw-p 00000000 00:0c 37798630                   /anon_hugepage (deleted)
```

**Comment by shefty** [comment]
The uffd thread is waiting on a lock acquired by the mlx5 library? Are there other threads running?  Can you determine which thread is holding that lock?

**Comment by jswaro** [comment]
Two threads in total. The uffd thread is waiting on a lock in glibc's free function is how I'm interpretting it at the moment. The main application thread is also in a call to the free function, but I can see that it is stuck waiting for a completion from user space. I am assuming that the main thread is waiting for the uffd thread to read the FD to issue the completion for kernel-space uffd, but the uffd thread is waiting on a lock that might be tied to the same region the main thread is operating on. Does that make sense so far?

**Comment by shefty** [comment]
See this line here:

https://github.com/linux-rdma/rdma-core/blob/master/providers/mlx5/verbs.c#L517

There is a free() call made in fi_dereg_mr().  The fi_dereg_mr() call can be made by the uffd thread itself as part of removing a cached entry that may have changed.  If that freed memory is being monitored as part of a larger region that is cached, I believe this will cause the hang you're seeing.  Free() hangs until another thread releases it.  That 'other' thread is supposed to be the uffd thread.

We need to change this, so that the uffd thread never tries to allocate or free memory.

**Comment by shefty** [comment]
@rajachan - tagging AWS, so that they are aware of this potential issue.

**Comment by jswaro** [comment]
> There is a free() call made in fi_dereg_mr(). The fi_dereg_mr() call can be made by the uffd thread itself as part of removing a cached entry that may have changed. If that freed memory is being monitored as part of a larger region that is cached, I believe this will cause the hang you're seeing. Free() hangs until another thread releases it. That 'other' thread is supposed to be the uffd thread.

I agree. To prevent the uffd from freeing memory, should some of these uffd event's be offloaded to EQ/CQ processing queues so that they are eventually processed by a non-uffd thread? Not ideal, but just thinking out lout.

**Comment by rajachan** [comment]
cc: @ofiwg/aws-efa-team

**Comment by jswaro** [comment]
@shefty Commenting out the notify for unmap allows the application to run to completion.

**Comment by shefty** [comment]
That has the potential of corrupting the application data.  You can try using the other memory monitor.

**Comment by jswaro** [comment]
> That has the potential of corrupting the application data. You can try using the other memory monitor.

I'm not suggesting it as a fix. I added the comment to support what you were saying above, and copied below. 

> If that freed memory is being monitored as part of a larger region that is cached, I believe this will cause the hang you're seeing. Free() hangs until another thread releases it.

**Comment by shefty** [comment]
@jswaro - ah, got it.  thanks.  I think we're going to need an additional queue/thread for processing notifications.  Maybe the EQ could work for that, but I'm not sure if that's the best approach to take, since it mixes EQ processing with internal event handling.  I wouldn't throw it out either, since it should avoid introducing an extra thread, and the queue already exists.

**Comment by jswaro** [comment]
I am contemplating the use of a secondary thread for the uffd handler, built around pthread_cond wait, and a freestack. It shouldn't be too disruptive since it will be sleeping until events are generated, and we'll end up reusing a fair amount of existing code from the util provider to make it. I think we need the additional thread to keep memory mappings with the device consistent as much as possible to prevent remote data from landing in stale registrations. If the event thread is in it's own thread, it should remove us from this deadlock that we are in at the moment. 

Thoughts?

**Comment by shefty** [comment]
I think that's a reasonable option.  The problem is making sure that we don't lose events, given that we can't have the uffd thread allocate memory when its freestack is empty.

**Comment by jswaro** [comment]
I was thinking we could put the uffd thread to sleep when it isn't able to enqueue the event into the freestack, and that would allow the event thread to process events and put them back into the freestack for use. We could signal the uffd thread, or simply hold the uffd.monitor.lock and allow it to be 'signaled' via mutex unlock. That could potentially avoid the nasty problem of memory allocation and deallocation from the uffd thread context.

**Comment by shefty** [comment]
We can't put the uffd thread to sleep or block it, or we can end up in the same situation.  If the event handling calls alloc/free, that can trigger another event, which the uffd thread must respond to.

**Comment by jswaro** [comment]
Ok. I'll give it some thought to how we can avoid the alloc/dealloc issue.

**Comment by shefty** [comment]
The monitor can still invoke the notification callback, and the callback can still perform the work of searching for cached registrations.  It will just need to defer any cleanup to another thread or work queue.  This should be able to avoid needing to allocate additional structures, since the cached registration entry can be used for that purpose.

As a proposal, we can try adding a flush_list to the mr_cache.  ofi_mr_cache_notify() would insert any matching cached entries onto the flush_list.  The flush_list would be processed at the start of all mr_cache calls (e.g. ofi_mr_cache_find, ofi_mr_cache_reg, ofi_mr_cache_delete, ofi_mr_cache_flush, etc.).

We need to be careful that we don't even end up calling free indirectly, for example, as part of removing a cached entry from the rbtree.

**Comment by jswaro** [comment]
I had given thought to an approach like that, but my concern relates to my earlier statement copied below. If the processing is deferred to an internal queue in the MR cache, it might not be processed until some time later unless we can get a new thread to interact with registered MR caches to push those events asynchonously of the uffd and application theads . 

>  I think we need the additional thread to keep memory mappings with the device consistent as much as possible to prevent remote data from landing in stale registrations.

**Comment by shefty** [comment]
If the local app frees memory while a remote peer is trying to RDMA into it, that's really an application level issue.  My bigger concern is that we cannot match an entry in the cache after the VA to PA might change.  That would result in data corruption that it outside the scope of the application.  To avoid that, we need the memory monitor to have most of the same behavior.

Specifically, the monitor must hold the cache lock while reading the event from the kernel.  It should continue to hold the lock while notifying the cache that a memory range should no longer be considered valid.  The cache must remove any matching entries to prevent future searches from matching against it.

There should not be any active RDMA operations incoming against that region.  Again, if there is, this is a bug in the upper layers/application.  It's okay for the rkey to continue to exist until garbage collection can free it.  No RDMA should be targeting it.

**Comment by jswaro** [comment]
> There should not be any active RDMA operations incoming against that region. Again, if there is, this is a bug in the upper layers/application. It's okay for the rkey to continue to exist until garbage collection can free it. No RDMA should be targeting it.

Agreed. As a personal preference, I would rather have an application observe a CQ error on a data transaction to a stale buffer instead of a 'successful' completion. It's a hard problem to solve correctly, and it's a race. Any attempt to defer the operation increases the likelihood of the stale data transfer scenario.

**Comment by jswaro** [comment]
> Specifically, the monitor must hold the cache lock while reading the event from the kernel. It should continue to hold the lock while notifying the cache that a memory range should no longer be considered valid.

A 'stale' flag in the registration struct should suffice right?

**Comment by shefty** [comment]
A stale flag might, but then the flag would need to be included as part of the sorting of entries.  This was why I suggested removing any found entries from the rbtree and storing them on a flush list.  Whether another thread processes the flush list or we do it on the next cache access shouldn't matter for correctness.

**Comment by jswaro** [comment]
I think making this work with the existing structures might require new functions which do no invoke free or malloc in the underlying data structures. rbtPrune or something similarly named that removes an entry from the tree without freeing the entry and leaving that to the application to manage?

**Comment by shefty** [comment]
A prune type call might work, or the rbtree can be updated so that the structure being inserted / removed provides the necessary storage for maintaining the tree (if it doesn't do that already).

**Comment by wckzhang** [comment]
@shefty Originally, I imagined that we could have a queue that ofi_monitor_notify would push to. We would handle events in this queue whenever we had a mr cache call. I thought this was in my original code, but maybe it got lost when it became util code. This matches your flush_list proposal pretty similarly. I think what we have to be careful of is how we hold the locks. For example, we can't hold the monitor lock while we parse this list, since we would then also have a deadlock. But if we don't hold this lock, the userfault thread will be triggered, receive a corresponding event, and then add an entry to the queue. Once we continue to the mr cache code and hold the lock, we will have unread entries in the queue.

**Comment by wckzhang** [comment]
The current design of the uffd_handler requires that uffd.monitor.lock be hold

**Comment by wckzhang** [comment]
Reading more comments...

I believe that anytime we call free within uffd.monitor.lock can result in a deadlock situation. The uffd thread won't be able to progress, and the faulting thread won't be able to wake up. This means that we either cannot protect our tree, or we risk a deadlock with only the monitor lock. We may need another lock, but will require more thinking.

**Comment by shefty** [comment]
I think we will need an additional lock around the cleanup list, which is unfortunate.

**Comment by jswaro** [comment]
I've started writing something to queue the events in the monitor and progress them on mr cache function entry.  We could still use another thread to flush the stale entries, but it wouldn't be strictly necessary.

**Comment by wckzhang** [comment]
One question I have is: Why is this memory being monitored? I don't see why the verbs_mr memory is being monitored. Are you registering the verbs_mr for some reason (I don't believe it can be a stale entry since the userfaultfd would catch it by the time it's malloc'd again)?

**Comment by jswaro** [comment]
I'm not sure that the mr memory is. Sean's explanation above suggested to me that you generally cannot call free or malloc from the uffd context.

**Comment by jswaro** [comment]
> I believe that anytime we call free within uffd.monitor.lock can result in a deadlock situation.

I don't think that is true.

If the free is within the context of the user application thread. If free occurs from the user thread, while holding the lock, uffd cannot read the events until we let the lock go. Once we let the lock go, uffd can read the event, and we push the event into a deferred queue in the mr cache -- avoiding any free or malloc calls in the process. 

The free cannot happen within the context of the uffd thread according to our discussions in this thread, so I won't go into that case. 

I'm going to put up a PR for the issue to open up the discussion on implementation. I think the lock may be unnecessary from the mr cache context since the monitor lock is always taken when operating on the mr cache in most meaningful ways. I believe I should be using a fastlock and I'll fix that when I get home. There are probably other modifications, so feel free to add to it.

**Comment by shefty** [comment]
From the viewpoint of the application or middleware, it is trying to send data over the fabric.  To do so requires that the memory be registered with the device.  However, the application has no concept of this requirement.  The result is that the provider registers the memory for the application (an expensive operation), then performs the transfer.  If the app issues another transfer that uses the same memory buffer, we save the registration cost on subsequent transfers by finding that the buffer has already been registered.

The issue here is that the application can free/unmap the memory buffer that has been cached.  If they allocate a new buffer, they can get the same virtual address, but now the physical pages backing the virtual address has changed.  The purpose of the monitor is to identify when the virtual to physical pages could have changed, so that the cached entry can be removed.  This way, a new transfer will be forced to register the memory (with the updated mappings) again.

We cannot call malloc/free/mmap/etc. from the uffd thread.  We also cannot hold the uffd lock while calling malloc/free/etc.  The latter is because the thread that calls malloc/free will hang in the kernel until the uffd thread reads the event from the userfault fd.  And in order to read that event, it must hold the uffd lock.

**Comment by shefty** [comment]
What we really need is a rbtree implementation that doesn't call free() as part of delete...

**Comment by wckzhang** [comment]
> We cannot call malloc/free/mmap/etc. from the uffd thread. We also cannot hold the uffd lock while calling malloc/free/etc. The latter is because the thread that calls malloc/free will hang in the kernel until the uffd thread reads the event from the userfault fd. And in order to read that event, it must hold the uffd lock.

>I'm not sure that the mr memory is. Sean's explanation above suggested to me that you generally cannot call free or malloc from the uffd context.

As far as I understand, it's preferable not to call malloc/free/mmap in the userfault thread to reduce risk, but it doesn't cause any issues if the memory being freed was not registered with the userfault fd. Memory that is not registered with the userfault fd won't even notify the poll function for userfault fd. Thus, this situation can only occur with memory that has been registered with the userfault fd being freed (Which the fd will immediately unregister from itself after we read from it).

**Comment by shefty** [comment]
It's true that as long as we haven't registered the memory being freed with the uffd, we're okay, but it's not easy to figure out if memory has been registered or not.  uffd registration requires aligning on page boundaries.  If a user's data transfer buffer shares the same page as some other structure (such as a mr_cache entry structure), then attempting to free that other structure can cause issues.  When huge pages are present, the chance of this happening seems high.

**Comment by wckzhang** [comment]
>  uffd registration requires aligning on page boundaries

As far as I understand, memory registrations pin pages, so they are technically page aligned as well.

> If a user's data transfer buffer shares the same page as some other structure (such as a mr_cache entry structure), then attempting to free that other structure can cause issues.

If a user's data transfer buffer shares the same page as another structure, calling free on that structure wouldn't cause an unmap event to fire off if the transfer buffer is still part of that page.

It may be possible that there is a situation where

1. Allocate two buffers, A & B which exist on the same page. A buffer C exists separately and is already registered
2. Register buffer A for sending, use buffer B for meta data.
3. Unregister buffer A and free buffer A - does not cause an unmap event yet since buffer B still exists. On unregistration we also don't unmonitor the page due to previous issues. 
4. Buffer C is unregistered and unmapped which triggers an unmap event.
5. Uffd handler is triggered and begins removing entries from the tree. It frees meta data structure C which is then unmapped and triggers another event, causing a deadlock.

> When huge pages are present, the chance of this happening seems high.

With huge pages, at least explicit huge pages, I thought you are required to use mmap and the buffer itself will be page aligned, so I don't this can happen with huge pages.

**Comment by shefty** [comment]
Registered buffers will pin the entire page, but the accessible region will start at the virtual address specified by the app.

A single physical page may back multiple malloc calls.  Take this scenario: app calls malloc, library calls malloc, app performs a transfer, app calls free, the library calls free.  This can result in an unmap event as a result of the library free call.

We can't predict what how the C library, or some other memory management library is behaving here.  And a registered region could start at any offset into a page.  I still think any call to malloc/free/etc. from the uffd thread or while holding the uffd lock has the potential to result in a hang.

One possible change we can make is to modify the rbtree to avoid calling free() from ofi_rbmap_delete().  We can instead push the obi_rbnode onto a free_list, and change ofi_rbmap_insert() to check the free_list first, before calling malloc.  ofi_rbmap_destroy() would need to cleanup the free_list.

**Comment by jswaro** [comment]
> It's true that as long as we haven't registered the memory being freed with the uffd, we're okay, but it's not easy to figure out if memory has been registered or not.



> As far as I understand, memory registrations pin pages, so they are technically page aligned as well.



> With huge pages, at least explicit huge pages, I thought you are required to use mmap and the buffer itself will be page aligned, so I don't this can happen with huge pages.



> We can't predict what how the C library, or some other memory management library is behaving here.

Generally speaking, we probably should avoid memory management function calls entirely within the context of the uffd thread. It is going to be difficult to handle these error cases without overcomplicating the code. Even with assumptions we can make about some systems, such as explicit huge pages, we run the risk of failure with other systems such as transparent huge pages. 



> One possible change we can make is to modify the rbtree to avoid calling free() from ofi_rbmap_delete()

It could help, but it wouldn't fix the problem entirely. the deregistration function may end up calling free as well.

**Comment by shefty** [comment]
Correct, the deregistration must still be deferred.  I'll update rbmap to remove the free from the delete call.  That's needed, so we can safely remove entries from the cache from within the uffd notification callback.  The entries can then be moved to a flush list for deferred cleanup.

As for the locking, we can make this work with a single uffd lock.  The general proposal would follow a flow like this:

```
ofi_mr_cache_notify(...)
{
    ....
    for every entry that overlaps with the reported region {
        remove entry from rbtree
        insert entry on cleanup list
    }
}

ofi_mr_cache_search(...)  /* or other exposed APIs */
{
    lock(monitor)
    while (cleanup list is not empty) {
        entry = remove head of cleanup list
        unlock(monitor)
        destroy entry /* unregister / free */
        lock(monitor)
    }

    ...
}
```

We need to adjust ofi_mr_cache_flush to release the monitor lock in a similar fashion.

**Comment by jswaro** [comment]
> We need to adjust ofi_mr_cache_flush to release the monitor lock in a similar fashion.

I'll get that PR up for general review. It is not complete though.

**Comment by wckzhang** [comment]
That proposal looks okay to me.

**Comment by jswaro** [comment]
> That proposal looks okay to me.

Fine to me as well.

**Comment by jswaro** [comment]
This is going to require some modification to the rbtree functions. The search function runs indefinitely, finding the same entry over and over again. This occurs even if you add logic to the search function for checking to see if the entry has been marked stale. To me, this means we need a function that returns the memory to be freed later to the upper layer, so that we can mark the entry stale, and pull it out of the tree.

**Comment by shefty** [comment]
I have the rbtree update.  Fixing up the cache flushing / cleanup isn't too hard, but will take some code unraveling.

The bigger issue is dealing with registration when we have a cache miss, and being able to do that without holding the monitor lock in a way that's also not racy...

**Comment by jswaro** [comment]
Can you post it to my PR, or send me a diff? I've been working on an rbtree change, so I'll drop my changes and pull yours in.

**Comment by shefty** [comment]
I'm pushing changes as I make them to: #5562 .  I *think* the next step in that series would be to restructure ofi_mr_cache_flush (and the underlying calls), to flush the lru_list without holding the monitor lock.

What I would like to do is have the flush_list be cleaned up as part of ofi_mr_cache_flush (or some subcall), and call that function for deferred cleanup.  I'm still working out how to best structure the code for this.

The bigger problem that I'm seeing is how to deal with memory registration being done while holding the monitor lock.  Trying to move that outside of the lock results in races with the uffd thread reporting events.  I have no proposed solution for that yet.

**Comment by shefty** [comment]
Fixed by #5562
